tests/test_model_system.py

```diff
@@ -439,3 +441,75 @@ def test_normalize(self):
         assert np.isclose(model_system.elemental_composition[0].atomic_fraction, 2 / 3)
         assert model_system.elemental_composition[1].element == 'O'
         assert np.isclose(model_system.elemental_composition[1].atomic_fraction, 1 / 3)
+
+
+    @pytest.mark.parametrize(
+        'mol_label_list, n_mol_list, atom_labels_list, composition_formula_list',
```

Comment:
I wondered whether you can test this a bit differently, if you decide to move everything to its own method class `resolve_composition_formula`. In that case, you can control much more when to return none, and when to return the composition formula.

If so, the testing should change to handle this "false testing", i.e., that when the conditions you are looking for do not apply, then the method returns None.

An example (this params are for seeing whether normalization of `KSpace` would return None, compute, or return the reciprocal_lattice_vectors): https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/blob/develop/tests/test_numerical_settings.py#L51

Reply:
I definitely agree that testing can be improved, thanks for the tips.

I am a bit unsure atm how to proceed, because I do not know how to deal with the recursive setting of the formula and the idea of returning the formula at each level instead of setting it. This is what I wrote above about it but not sure if you saw:
"""
I was assuming that only branch_depth = 0 would be ever marked as representative. That's why I am traversing the entire tree to set the composition formula at each branch level.

I could instead resolve the composition formula for all model systems, representative or not, then I don't need to have any recursive function, I just look at the children of the current branch. The difficulty there though is that for the bottom of the hierarchy I need to go back up the tree to find the list of atom labels
"""

Do you think I should normalize only for representative systems and traverse the tree downwards, or normalize all systems, but have to traverse to the top of the tree for any of the bottom nodes?

Reply:
Ok, I moved the normalization up to Simulation() as you suggested. I think this is nicer. I need to still work on the tests though. I will have to come back to that next week.

Reply:
Ok, cool. Let me know when you need another review. I am resolving all conversations 👍🏻